src/pm/hydra/proxy/pmip_cb.c

```diff
@@ -458,22 +458,214 @@ static HYD_status pmi_listen_cb(int fd, HYD_event_t events, void *userp)
+    HYDU_ERR_POP(status, "error reading command from launcher\n");
+    HYDU_ASSERT(!closed, status);
+
+    if (hdr.cmd == CMD_PROC_INFO) {
+        status = handle_launch_procs(fd);
+        HYDU_ERR_POP(status, "launch_procs returned error\n");
+    } else if (hdr.cmd == CMD_PMI_RESPONSE) {
+        status = handle_pmi_response(fd, hdr.buflen, hdr.u.pmi.pmi_version, hdr.u.pmi.process_fd);
+        HYDU_ERR_POP(status, "unable to handle PMI response\n");
+    } else if (hdr.cmd == CMD_SIGNAL) {
+        int signum = hdr.u.data;
+        /* FIXME: This code needs to change from sending the signal to
+         * a PMI-2 notification message. */
+        HYD_pmcd_pmip_send_signal(signum);
+    } else if (hdr.cmd == CMD_STDIN) {
+        int count;
+
+        if (hdr.buflen) {
+            HYDU_MALLOC_OR_JUMP(buf, char *, hdr.buflen, status);
+            HYDU_ERR_POP(status, "unable to allocate memory\n");
+
+            status = HYDU_sock_read(fd, buf, hdr.buflen, &count, &closed, HYDU_SOCK_COMM_MSGWAIT);
+            HYDU_ERR_POP(status, "unable to read from control socket\n");
+            HYDU_ASSERT(!closed, status);
+
+            if (HYD_pmcd_pmip.downstream.in == HYD_FD_CLOSED) {
+                MPL_free(buf);
+                goto fn_exit;
+            }
+
+            status = HYDU_sock_write(HYD_pmcd_pmip.downstream.in, buf, hdr.buflen, &count,
+                                     &closed, HYDU_SOCK_COMM_NONE);
+            HYDU_ERR_POP(status, "unable to write to downstream stdin\n");
+
+            HYDU_ERR_CHKANDJUMP(status, count != hdr.buflen, HYD_INTERNAL_ERROR,
+                                "process reading stdin too slowly; can't keep up\n");
+
+            HYDU_ASSERT(count == hdr.buflen, status);
+
+            if (HYD_pmcd_pmip.user_global.auto_cleanup) {
+                HYDU_ASSERT(!closed, status);
+            } else if (closed) {
+                close(HYD_pmcd_pmip.downstream.in);
+                HYD_pmcd_pmip.downstream.in = HYD_FD_CLOSED;
+            }
+
+            MPL_free(buf);
+        } else {
+            close(HYD_pmcd_pmip.downstream.in);
+            HYD_pmcd_pmip.downstream.in = HYD_FD_CLOSED;
+        }
+    } else {
+        status = HYD_INTERNAL_ERROR;
     }
 
-    return ret;
+    HYDU_ERR_POP(status, "error handling proxy command\n");
+
+  fn_exit:
+    HYDU_FUNC_EXIT();
+    return status;
+
+  fn_fail:
+    goto fn_exit;
+}
+
+/* handle_launch_procs - read proc_info and launch procs */
+
+static HYD_status parse_exec_params(char **t_argv);
+static HYD_status procinfo(int fd);
+static HYD_status singleton_init(void);
+static HYD_status launch_procs(void);
+static int local_to_global_id(int local_id);
+
+static HYD_status handle_launch_procs(int fd)
+{
+    HYD_status status = HYD_SUCCESS;
+    HYDU_FUNC_ENTER();
+
+    status = procinfo(fd);
+    HYDU_ERR_POP(status, "error parsing process info\n");
```

Comment:
Process binding initialization should be here when merge conflicts are fixed.

Reply:
Good reminder! I have rebased the PR.